File: python/linked_list/linked_list_imple.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self, value):
        self.head = Node(value)

    def append(self, value):
        cur = self.head

        while cur.next is not None:
            cur = cur.next
        cur.next = Node(value)

    def print_all(self):
        cur = self.head

        while cur is not None:
            print(cur.data)
            cur = cur.next

    def get_node(self, index):
        cur = self.head
        cur_index = 0

        while cur_index != index:
            cur = cur.next
            cur_index += 1

        return cur

    def add_node(self, index, value):
        # index 위치에 value 추가하기
        new_node = Node(value)

        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return

        pre_node = self.get_node(index - 1)
        next_node = pre_node.next
        pre_node.next = new_node
        new_node.next = next_node

    def delete_node(self, index):
        if index == 0:
            self.head = self.head.next
            return

        pre_node = self.get_node(index - 1)
        pre_node.next = self.get_node(index).next


def get_single_linked_list_sum(li):
    sum = 0
    cur = li.head

    while cur is not None:
        sum = sum * 10 + cur.data
        cur = cur.next

    return sum


def get_linked_list_sum(linked_list_1, linked_list_2):
    sum_1 = get_single_linked_list_sum(linked_list_1)
    sum_2 = get_single_linked_list_sum(linked_list_2)

    return sum_1 + sum_2


if __name__ == "__main__":
    # l = LinkedList(1)
    # l.append(2)
    # l.append(3)
    # l.add_node(1, 3)
    # l.add_node(0, 7)
    #
    # l.delete_node(0)
    #
    # l.print_all()

    # print("1번째 node index = ", l.get_node(1).data)

    # 문제1 - 두개의 링크드리스트에 있는 값들의 더한 값을 구하시오
    # 6 / 6 * 10 + 7 = 67 / 67 * 10 + 8 = 678
    l1 = LinkedList(6)
    l1.append(7)
    l1.append(8)

    l2 = LinkedList(3)
    l2.append(5)
    l2.append(4)

    print(get_linked_list_sum(l1, l2))

File: java/src/linkedList/LinkedListImple.java
package linkedList;

public class LinkedListImple {
    public static void main(String[] args) {
//        LinkedListEx l = new LinkedListEx(1);
//        l.append(7);
//        l.append(5);
//        l.append(2);
//        l.append(3);
//
//        l.addNode(1, 2);
//        l.addNode(0, 9);
////        l.deleteNode(0);
//
//        l.printAll();

//        System.out.println(l.getNode(1).data);

        LinkedListEx li1 = new LinkedListEx(6);
        li1.append(7);
        li1.append(8);

        LinkedListEx li2 = new LinkedListEx(3);
        li2.append(5);
        li2.append(4);

        System.out.println(getLinkedListSum(li1, li2));
    }

    static int getLinkedListSum(LinkedListEx linkedList1, LinkedListEx linkedList2) {
        int sum1 = getSingleListSum(linkedList1);
        int sum2 = getSingleListSum(linkedList2);

        return sum1 + sum2;
    }

    static int getSingleListSum(LinkedListEx linkedList) {
        int sum = 0;
        Node cur = linkedList.head;

        while (cur != null) {
            sum = sum * 10 + cur.data;
            cur = cur.next;
        }

        return sum;
    }
}

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedListEx {
    Node head;

    public LinkedListEx(int value) {
        this.head = new Node(value);
    }

    public void append(int value) {
        Node cur = head;

        while (cur.next != null) {
            cur = cur.next;
        }
        cur.next = new Node(value);
    }

    public void printAll() {
        Node cur = head;

        while (cur != null) {
            System.out.println(cur.data);
            cur = cur.next;
        }
    }

    public Node getNode(int index) {
        Node cur = head;
        int curIndex = 0;

        while (curIndex != index) {
            if (cur == null) {
                throw new IndexOutOfBoundsException("Index " + index + " out of range");
            }
            cur = cur.next;
            curIndex++;
        }

        return cur;
    }

    public void addNode(int index, int value) {
        Node newNode = new Node(value);

        if (index == 0) {
            newNode.next = head;
            head = newNode;
            return;
        }

        Node preNode = getNode(index - 1);
        Node nextNode = preNode.next;
        preNode.next = newNode;
        newNode.next = nextNode;
    }

    public void deleteNode(int index) {
        if (index == 0) {
            head = head.next;
            return;
        }

        Node preNode = getNode(index - 1);
        Node deleteNode = getNode(index);
        preNode.next = deleteNode.next;
    }
}
